src/sos/objectstorage.go

```diff
@@ -0,0 +1,124 @@
+	"golang.org/x/exp/slices"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+var (
+	organizationLabel = "appuio.io/organization"
+	namespaceLabel    = "crossplane.io/claim-namespace"
+)
+
+type ObjectStorage struct {
+	k8sClusters    *[]kubernetes.Cluster
+	exoscaleClient *egoscale.Client
+	bucketDetails  []BucketDetail
+}
+
+type BucketDetail struct {
+	ClusterName  string
+	Namespace    string
+	Organization string
+	BucketName   string
+}
+
+func NewObjectStorage(exoscaleClient *egoscale.Client, k8sClusters *[]kubernetes.Cluster) ObjectStorage {
+	return ObjectStorage{
+		exoscaleClient: exoscaleClient,
+		k8sClusters:    k8sClusters,
+	}
+}
+
+func (o *ObjectStorage) Execute(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Running metrics collector by step")
+
+	p := pipeline.NewPipeline[context.Context]()
+	p.WithSteps(
+		p.NewStep("Fetch managed buckets", o.fetchManagedBuckets),
+		p.NewStep("Get bucket usage", o.getBucketUsage),
+	)
+	return p.RunWithContext(ctx)
+}
+
+func (o *ObjectStorage) getBucketUsage(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Fetching bucket usage from Exoscale")
+	resp, err := o.exoscaleClient.ListSosBucketsUsageWithResponse(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, bucketUsage := range *resp.JSON200.SosBucketsUsage {
+		name := *bucketUsage.Name
+		zoneName := *bucketUsage.ZoneName
+		size := *bucketUsage.Size
+		createdAt := *bucketUsage.CreatedAt
+
+		log.V(1).Info("Trying to find a match in a cluster", "bucket name", name)
+		// Match bucket name from exoscale with the bucket found in one of the clusters
+		bucketIndex := slices.IndexFunc(o.bucketDetails, func(ni BucketDetail) bool { return ni.BucketName == name })
+
+		if bucketIndex == -1 {
+			log.Info("Cannot find bucket in any cluster", "bucket name", name)
+			continue
+		}
+		fmt.Printf("name: %s, zoneName: %s, size: %d, createdAt: %s, namespace info: %s\n", name, zoneName, size, createdAt, o.bucketDetails[bucketIndex])
+	}
+	return nil
+}
+
+func (o *ObjectStorage) fetchManagedBuckets(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Fetching buckets from clusters")
+	for _, k8sCluster := range *o.k8sClusters {
+		buckets := exoscalev1.BucketList{}
+		log.V(1).Info("Listing buckets from cluster", "cluster name", k8sCluster.Name)
+		err := k8sCluster.List(ctx, &buckets)
+		if err != nil {
+			return fmt.Errorf("cannot list buckets: %w", err)
+		}
+		bucketDetailsByCluster, err := matchOrgAndNamespaceByBucket(ctx, &buckets, k8sCluster.Name)
+		if err != nil {
+			return fmt.Errorf("cannot fetch bucket details from buckets: %w", err)
+		}
+		// add BucketDetail from each cluster to one single slice
+		o.bucketDetails = append(o.bucketDetails, *bucketDetailsByCluster...)
+	}
+	return nil
+}
+
+func matchOrgAndNamespaceByBucket(ctx context.Context, buckets *exoscalev1.BucketList, clusterName string) (*[]BucketDetail, error) {
```

Comment:
This function has an error return in the signature, but it only has one single return with a nil error. So returning an error doesn't make sense.

Also, the function adds the buckets to the slice even though they might be missing information. Not sure if that makes sense. We should probably just skip them, we can't really use them anyway in later stages.

Reply:
Had a call and resolved these points.

Reply:
As discussed I will move the logs when I implement the database integration.